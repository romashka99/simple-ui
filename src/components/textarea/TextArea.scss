@import '../../styles/colors.scss';
@import '../../styles/main.scss';
@import '../../styles/style.scss';

.textarea {
	display: flex;
	flex-direction: column;

	& .textarea--label,
	& .textarea--help {
		font-size: 12px;
		line-height: 16px;
	}

	& .textarea--help {
		display: flex;
		justify-content: space-between;
	}

	&.is-disabled {
		.textarea--label,
		.textarea--help {
			color: $color-mid-light-grey;
		}
	}

	.textarea--inner {
		padding: 12px 8px;
		border: $border;
		border-radius: $border-radius;
		background-color: $color-white;
		color: $color-grey;
		line-height: 24px;
		font-size: 16px;
		box-sizing: border-box;

		&::placeholder {
			color: $color-grey;
		}

		&:disabled {
			border-color: $color-mid-light-grey;
			color: $color-mid-light-grey;
			cursor: not-allowed;

			&::placeholder {
				color: $color-mid-light-grey;
			}
		}

		&:not(:disabled) {
			&:hover {
			}
			&:active,
			&:focus {
			}
		}
	}
}
