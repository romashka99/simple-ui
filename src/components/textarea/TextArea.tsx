import React, {
	ChangeEvent,
	DetailedHTMLProps,
	TextareaHTMLAttributes,
	useCallback,
	useState,
} from 'react';
import './TextArea.scss';

interface ITextAreaProps
	extends DetailedHTMLProps<
			TextareaHTMLAttributes<HTMLTextAreaElement>,
			HTMLTextAreaElement
		>,
		React.AriaAttributes {
	label?: string;
	helpText?: string;
	showCount?: boolean;
}

const getCount = (
	value?: string | number | readonly string[],
	maxLength?: number
) => {
	return `${
		value
			? typeof value !== 'number'
				? value?.length
				: value.toString().length
			: 0
	}/${maxLength}`;
};

export default function TextArea({
	label,
	helpText,
	showCount = false,
	value,
	maxLength,
	disabled,
	onChange,
	className,
	...prop
}: ITextAreaProps) {
	const classes = `textarea${disabled ? ' is-disabled' : ''}${
		className ? ' ' + className : ''
	}`;

	const [valueTextarea, setValueTextarea] =
		useState<string | number | readonly string[] | undefined>(value);

	const onChangeValue = useCallback(
		(event: ChangeEvent<HTMLTextAreaElement>) => {
			const { value } = event.target;
			setValueTextarea(value);
			if (onChange) onChange(event);
		},
		[onChange]
	);

	return (
		<label data-testid='textarea' className={classes}>
			{label && <span className='textarea--label'>{label}</span>}
			<textarea
				{...prop}
				data-testid='textarea-inner'
				value={valueTextarea}
				maxLength={maxLength}
				className='textarea--inner'
				disabled={disabled}
				onChange={onChangeValue}
			/>
			{(helpText || (showCount && maxLength)) && (
				<div className='textarea--help'>
					{helpText && <div>{helpText}</div>}
					{showCount && maxLength && (
						<div>{getCount(valueTextarea, maxLength)}</div>
					)}
				</div>
			)}
		</label>
	);
}
