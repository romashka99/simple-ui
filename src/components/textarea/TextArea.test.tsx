import { render, screen, within } from '@testing-library/react';
import React from 'react';
import TextArea from '.';
import userEvent from '@testing-library/user-event';

describe('Test TextArea', () => {
	test('without attr', () => {
		render(<TextArea />);
		const textarea = screen.getByTestId('textarea');
		const inner: HTMLInputElement =
			within(textarea).getByTestId('textarea-inner');
		expect(inner.disabled).toEqual(false);
		expect(textarea.classList.length).toBe(1);
		expect(textarea.classList.contains('textarea')).toBe(true);
		expect(textarea).toMatchSnapshot();
	});

	test('has label', () => {
		render(<TextArea label='Label' />);
		const textarea = screen.getByTestId('textarea');
		const label = within(textarea).getByText('Label');
		expect(label).not.toBeNull();
		expect(textarea).toMatchSnapshot();
	});

	test('has help text', () => {
		render(<TextArea helpText='Help text' />);
		const textarea = screen.getByTestId('textarea');
		const helpText = within(textarea).getByText('Help text');
		expect(helpText).not.toBeNull();
		expect(textarea).toMatchSnapshot();
	});

	test('disabled', () => {
		render(<TextArea disabled />);
		const textarea = screen.getByTestId('textarea');
		const inner: HTMLInputElement =
			within(textarea).getByTestId('textarea-inner');
		expect(inner.disabled).toEqual(true);
		expect(inner.value).toEqual('');
		userEvent.type(inner, 'abc');
		expect(inner.value).toEqual('');
		expect(textarea).toMatchSnapshot();
	});

	test('readOnly', () => {
		render(<TextArea readOnly />);
		const textarea = screen.getByTestId('textarea');
		const inner: HTMLInputElement =
			within(textarea).getByTestId('textarea-inner');
		expect(inner.readOnly).toEqual(true);
		expect(inner.value).toEqual('');
		userEvent.type(inner, 'abc');
		expect(inner.value).toEqual('');
		expect(textarea).toMatchSnapshot();
	});

	test('onChange', () => {
		let value = 0;
		render(
			<TextArea
				onChange={() => {
					value = 1;
				}}
			/>
		);
		const textarea = screen.getByTestId('textarea');
		const inner: HTMLInputElement =
			within(textarea).getByTestId('textarea-inner');
		expect(value).toBe(0);
		userEvent.type(inner, 'abc');
		expect(value).toBe(1);
		expect(inner.value).toBe('abc');
		expect(textarea).toMatchSnapshot();
	});
});
