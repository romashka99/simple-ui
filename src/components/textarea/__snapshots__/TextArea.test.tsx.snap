// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TextArea disabled 1`] = `
<label
  class="textarea is-disabled"
  data-testid="textarea"
>
  <textarea
    class="textarea--inner"
    data-testid="textarea-inner"
    disabled=""
  />
</label>
`;

exports[`Test TextArea has help text 1`] = `
<label
  class="textarea"
  data-testid="textarea"
>
  <textarea
    class="textarea--inner"
    data-testid="textarea-inner"
  />
  <div
    class="textarea--help"
  >
    <div>
      Help text
    </div>
  </div>
</label>
`;

exports[`Test TextArea has label 1`] = `
<label
  class="textarea"
  data-testid="textarea"
>
  <span
    class="textarea--label"
  >
    Label
  </span>
  <textarea
    class="textarea--inner"
    data-testid="textarea-inner"
  />
</label>
`;

exports[`Test TextArea onChange 1`] = `
<label
  class="textarea"
  data-testid="textarea"
>
  <textarea
    class="textarea--inner"
    data-testid="textarea-inner"
  >
    abc
  </textarea>
</label>
`;

exports[`Test TextArea readOnly 1`] = `
<label
  class="textarea"
  data-testid="textarea"
>
  <textarea
    class="textarea--inner"
    data-testid="textarea-inner"
    readonly=""
  />
</label>
`;

exports[`Test TextArea without attr 1`] = `
<label
  class="textarea"
  data-testid="textarea"
>
  <textarea
    class="textarea--inner"
    data-testid="textarea-inner"
  />
</label>
`;
